### ISSUE TITLE
No easy way to access score when using a ValuesSourceMetricsAggregationBuilder-based aggregation

### ISSUE THREAD

**Issue by udoprog** [elastic/elasticsearch#10091]
I attempted to implement the aggregation documented at the bottom of [this article](http://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-top-hits-aggregation.html). Since dynamic scripting is disabled by default, I was not able to find an easy way to implement it.

The solutions I've discovered so far are:

* Enable dynamic scripting via `script.disable_dynamic: false`
* Upload a stored script and use with `script_file`

All of which can be difficult if the developer using elasticsearch does not have access to administrate the service, or if it is a public facing service.

**Comment by theQuazz** [comment]
:+1:

I have tried using a Lucene Expressions to solve this problem, but it look like the language doesn't support this use case yet

**Comment by dakrone** [comment]
@rjernst how hard would it be to add super simple support for `_score` to expressions?

**Comment by rjernst** [comment]
I investigated this for a while.  Expressions already support scores; the problems are with aggs.

The first problem is scorers are not passed through the many layers of objects consistently.  In the case of the `max` agg example there, I think there are at least 2 levels missing, first in the anonymous `LeafBucketCollectorBase`, and then in the anonymous `NumericDoubleValues` created in `MultiValueMode.MAX.select()`.

The bigger problem though is possibly more of a design problem that might prevent this from working. Within `getLeafCollector` of the `MaxAggregator`, `doubleValues()` is called before creating the `LeafBucketCollectorBase`, which binds the values object to a specific leaf reader (ie segment). The collector there is what implements `ScorerAware`.  However, expressions, in order to be efficient, map the bindings to concrete function values eagerly (when the binding to a leaf reader occurs). Here it looks like a lazy binding, which probably works fine for `groovy` since its variables are a hash based lookup, but not for expressions.  This eager binding is what allows expressions to be really fast.

It is possible though I misunderstood the aggs classes. @jpountz do you have any thoughts here?

**Comment by egpbos** [comment]
+1

**Comment by clintongormley** [comment]
@jpountz you may have missed this?

**Comment by jpountz** [comment]
@clintongormley Yes totally!

I just tried on master with the following sense recreation and this now works:

```
DELETE test

PUT test
{
  "mappings": {
    "test": {
      "properties": {
        "cat": {
          "type": "string",
          "index": "not_analyzed"
        },
        "text": {
          "type": "string"
        }
      }
    }
  }
}

PUT test/test/1
{
  "cat": "a",
  "text": "foo bar"
}

PUT test/test/2
{
  "cat": "a",
  "text": "bar"
}

PUT test/test/3
{
  "cat": "b",
  "text": "bar quux"
}

PUT test/test/4
{
  "cat": "c",
  "text": "bar"
}

GET test/_search
{
  "query": {
    "match": {
      "text": "bar"
    }
  }, 
  "aggs": {
    "by_cat": {
      "terms": {
        "field": "cat",
        "size": 10,
        "order": {
          "top_score": "desc"
        }
      },
      "aggs": {
        "top_hits": {
          "top_hits": {}
        },
        "top_score": {
          "max": {
            "script": "_score",
            "lang": "expression"
          }
        }
      }
    }
  }
}
```

It apparently got fixed through https://github.com/elastic/elasticsearch/pull/10389/files#diff-867bb5fa33f1c72de5e99bc120550a7cR92. When setScorer is called on the LeafSearchScript object we re-create the function values in order to take the new scorer into account.

**Comment by avdv** [comment]
Is this still supported?

My aggregations look almost identical:

```
{
  "aggregations": {
    "parent": {
        "terms": {
            "field": "container",
            "size": 10,
            "order": {
                "max_score": "desc"
            }
        },
        "aggregations": {
            "max_score": {
                "max": {
                    "lang": "expression",
                    "script": "_score"
                }
            }
        }
    }
  }
}
```

but it fails with:

```
Caused by: java.lang.IllegalStateException: Expressions referencing the score can only be used for sorting
```

I'm using Elasticsearch 1.7.5.

**Comment by jpountz** [comment]
@avdv This is fixed in elasticsearch 2.0+.

**Comment by avdv** [comment]
@jpountz Thanks, it's time to upgrade I guess.. :wink:
